# Természetes nyelvfeldolgozás
# PTE Gépi tanulás III.

## 6. Gyakorlat
### 2024. március 13.


#Feladat

Töltsük le a telex (https://telex.hu/english?oldal=1) angol nyelvű oldaláról az első 10 cikket egy dataframe-be.


| Title                                             | Link                                                                                   | Text                                                      |
|---------------------------------------------------|----------------------------------------------------------------------------------------|-----------------------------------------------------------|
| Hungarian FM summons US ambassador over Biden’s ‘lies’ | [Link](https://telex.hu/english/2024/03/12/hungarian-fm-summons-us-ambassador-over-bidens-lies) | Hungary doesn't have to tolerate lies, even f...         |
| Hungarian government only one to vote against ... | [Link](https://telex.hu/english/2024/03/12/hungarian-government-only-one-to-vote-against)   | Member states voted on a regulation on transpa...       |
| Trump-admiring fake Twitter profile delivers p... | [Link](https://telex.hu/english/2024/03/12/trump-admiring-fake-twitter-profile-delivers-pro-russian-propaganda) | The perfect case for the newly established Sov...       |
| Orbán: Trump is a legend                          | [Link](https://telex.hu/english/2024/03/11/orban-trump-is-a-legend)                        | "He is indeed the seventh, which is a dangerou...       |
| Set decorator Zsuzsa Mihalek wins Oscar           | [Link](https://telex.hu/english/2024/03/11/set-decorator-zsuzsa-mihalek-wins-oscar)        |                                                           |
| He's the boss, he's a great leader – Trump hos... | [Link](https://telex.hu/english/2024/03/09/hes-the-boss-hes-a-great-leader-trump-ho...)     | Donald Trump and a small group of his close ad...       |
| VSquare: Tusk and Fiala were shouting at Orbán... | [Link](https://telex.hu/english/2024/03/08/vsquare-tusk-and-fiala-were-shouting-at-orban)  | Prime Minister Viktor Orbán of Hungary (L), Pe...       |
| Majority holds Orbán responsible for clemency ... | [Link](https://telex.hu/english/2024/03/08/majority-holds-orban-responsible-for-clemency)   | A representative poll conducted by Republikon ...       |
| Orbán met with Trump's former senior adviser i... | [Link](https://telex.hu/english/2024/03/07/orban-met-with-trumps-former-senior-adviser-in-washington-dc) | Prime Minister Viktor Orbán spoke on Thursday ...       |
| UNESCO warns Hungary against implementing orig... | [Link](https://telex.hu/english/2024/03/07/unesco-warns-hungary-against-implementing-original-boundary-of-ferto-landscape-world-heritage-site) | The Fertő landscape cannot bear any more const...       |


A cikkek szövegét (text) foglaljuk össze max 80 token hosszúságúra egy summarized mezőbe a facebook/bart-large-cnn modell segítségével.



Az összefoglalás szöveg hangulatát értékeljük a cardiffnlp/twitter-roberta-base-sentiment-latest modell segítségével. A label és score értékeket is adjuk hozzá a dataframe-hez.


Az összefoglalt angolnyelvű szövegeket fordítsuk le magyar nyelvre.

Készítsük elő egy emailhez a datafamet, úgy hogy HTML kóddá alakítjuk az alábbi formátum szerint.


<html>
<head>
<title></title>
</head>
<body>
<h1>Releváns cikkek összefoglalója</h1>
<p>Ez az email a legfrissebb releváns cikkeket foglalja össze.</p>
<ul>

  <li>
    <h2>Hungarian FM summons US ambassador over Biden’s remarks about Orbán</h2>
    <p><a href="https://telex.hu/english/2024/03/12/hungarian-fm-summons-us-ambassador-over-bidens-remarks-about-orban">https://telex.hu/english/2024/03/12/hungarian-fm-summons-us-ambassador-over-bidens-remarks-about-orban</a></p>
    <p>US President Joe Biden criticised Hungarian Prime Minister Viktor Orbán. The Foreign Ministry has summoned the US Ambassador for an "urgent" meeting. The ambassador was asked to produce a quote from the prime minister, but no such statement was made. The US ambassador also said that the official US position is that "here in Hungary we are building a dictatorship"</p>
    <p>Joe Biden elnök bírálta Orbán Viktor miniszterelnököt. A Külügyminisztérium "sürgős" találkozóra hívta össze az amerikai nagykövetet. A nagykövetet felkérték, hogy idézzen a miniszterelnöktől, de nem tett ilyen nyilatkozatot. Az amerikai nagykövet azt is elmondta, hogy a hivatalos amerikai álláspont az, hogy "itt Magyarországon diktatúrát építünk"</p>
  </li>
  
  <li>
    <h2>Hungarian government only one to vote against EU law on transparency in political advertising</h2>
    <p><a href="https://telex.hu/english/2024/03/12/hungarian-government-only-one-to-vote-against-eu-law-on-transparency-in-political-advertising">https://telex.hu/english/2024/03/12/hungarian-government-only-one-to-vote-against-eu-law-on-transparency-in-political-advertising</a></p>
    <p>Member states voted on a regulation on transparency in political advertising in the EU Council on Monday. The Hungarian government's representative was the only one to vote against the regulation. A qualified majority (i.e. 15 out of 27 countries accounting for 65% of the EU's total population) was enough for the proposal to pass.</p>
    <p>A magyar kormány képviselője volt az egyetlen, aki a rendelet ellen szavazott. A minősített többség (az EU teljes lakosságának 65%-át kitevő 27 ország közül 15) elegendő volt a javaslat elfogadásához.</p>
  </li>
  
  <li>
    <h2>Trump-admiring fake Twitter profile delivers perfect case for Sovereignty Protection Authority</h2>
    <p><a href="https://telex.hu/english/2024/03/12/trump-admiring-fake-twitter-profile-delivers-perfect-case-for-sovereignty-protection-authority">https://telex.hu/english/2024/03/12/trump-admiring-fake-twitter-profile-delivers-perfect-case-for-sovereignty-protection-authority</a></p>
    <p>The government-affiliated press in Hungary has been announcing in unison that there is now definite proof that George Soros funded the Hungarian opposition during the 2022 election campaign. All their articles are referencing videos published on an X (formerly Twitter) account dedicated to praising Donald Trump under the name MagaBabe. The videos feature author Kati Marton, chair of A4D's New York advisory board, and Dávid Korányi, the former advisor to Budapest mayor Gergely Karácsony. According to the government’s propaganda outlets the videos show them admitting</p>
    <p> A magyar kormányhoz tartozó sajtó egy időben bejelentette, hogy a 2022-es választási kampány során Soros György támogatta a magyar ellenzéket. Minden cikkük egy X-en (korábban Twitteren) közzétett videókra utal, amelyek Donald Trumpot MagaBabe néven dicsérik. A videók szerzője, Kati Marton, az A4D New York tanácsadó testületének elnöke, valamint Korányi Dávid, Karácsony Gergely volt tanácsadója.</p>
  </li>
  
  <li>
    <h2>Orbán: Trump is a legend</h2>
    <p><a href="https://telex.hu/english/2024/03/11/orban-trump-is-a-legend">https://telex.hu/english/2024/03/11/orban-trump-is-a-legend</a></p>
    <p>Viktor Orbán said he had worked with all seven Hungarian heads of state as PM and therefore had insight into "what kind of president was suitable for each situation" The Prime Minister said that if he becomes president again, Trump will not want to fund the security of Europeans nor will he support the US giving money to Ukraine.</p>
    <p>Viktor Orbán azt mondta, miniszterelnökként mind a hét magyar államfővel dolgozott együtt, ezért bepillantást nyert "milyen elnök alkalmas minden helyzetre" A miniszterelnök azt mondta, hogy ha újra elnök lesz, Trump nem fogja támogatni az európaiak biztonságát, és nem fogja támogatni az USA-t sem, hogy pénzt adjon Ukrajnának.</p>
  </li>
  
  <li>
    <h2>Set decorator Zsuzsa Mihalek wins Oscar</h2>
    <p><a href="https://telex.hu/english/2024/03/11/set-decorator-zsuzsa-mihalek-wins-oscar-for-her-work-on-poor-things">https://telex.hu/english/2024/03/11/set-decorator-zsuzsa-mihalek-wins-oscar-for-her-work-on-poor-things</a></p>
    <p>CNN.com will feature iReporter photos in a weekly Travel Snapshots gallery. Please submit your best shots of the U.S. for next week. Visit CNN.com/Travel next Wednesday for a new gallery of snapshots. Please share your best photos of the United States with CNN iReport.</p>
    <p> CNN.com egy heti Travel Snapshots galéria iReporter képekkel. Kérjük, küldje el a legjobb felvételeket az Egyesült Államok a jövő héten. Látogasson el a CNN.com/Utazz jövő szerdán egy új galéria a fényképek. Kérjük, ossza meg a legjobb fotókat az Egyesült Államok a CNN iReport.</p>
  </li>
  
  <li>
    <h2>He's the boss, he's a great leader – Trump hosts Orbán in friendly meeting at his Florida estate</h2>
    <p><a href="https://telex.hu/english/2024/03/09/hes-the-boss-hes-a-great-leader-trump-hosts-orban-in-friendly-meeting-at-his-florida-estate">https://telex.hu/english/2024/03/09/hes-the-boss-hes-a-great-leader-trump-hosts-orban-in-friendly-meeting-at-his-florida-estate</a></p>
    <p>Viktor Orbán and Donald Trump met for about an hour at the former president's estate in Mar-a-Lago, Florida. CNN sources described the meeting as a friendly chat without an agenda. Orbán has traditionally had a tense relationship with Democratic-led US administrations.</p>
    <p> Viktor Orbán és Donald Trump körülbelül egy órán át találkoztak a floridai Mar-a-Lago-ban lévő korábbi elnök birtokán. A CNN forrásai szerint a találkozó egy baráti beszélgetés volt napirend nélkül. Orbán hagyományosan feszült kapcsolatban állt a demokrata kormányokkal.</p>
  </li>
  
  <li>
    <h2>VSquare: Tusk and Fiala were shouting at Orbán while Fico remained silent</h2>
    <p><a href="https://telex.hu/english/2024/03/08/vsquare-tusk-and-fiala-were-shouting-at-orban-while-fico-remained-silent">https://telex.hu/english/2024/03/08/vsquare-tusk-and-fiala-were-shouting-at-orban-while-fico-remained-silent</a></p>
    <p>Viktor Orbán of Hungary, Petr Fiala of the Czech Republic, Donald Tusk of Poland and Robert Fico of Slovakia attended the V4 summit in Prague on 27 February 2024. Prime Minister of Hungary Viktor Orbán attended the summit with President of Czech Republic Petr Fischer and Prime Ministers of Poland, Czech Republic and Slovakia.</p>
    <p> Orbán Viktor, a Cseh Köztársaság Petr Fiala, a lengyel Donald Tusk és a szlovák Robert Fico részt vett a Prágai V4-csúcstalálkozón 2024. február 27-én.</p>
  </li>
  
  <li>
    <h2>Majority holds Orbán responsible for clemency decision, survey shows</h2>
    <p><a href="https://telex.hu/english/2024/03/08/majority-holds-orban-responsible-for-clemency-decision-survey-shows">https://telex.hu/english/2024/03/08/majority-holds-orban-responsible-for-clemency-decision-survey-shows</a></p>
    <p>Only one in four people are of the opinion that President Katalin Novák, who granted a presidential pardon to Endre K., is responsible. Opposition and governing party voters have very different views on the role of the President and the Prime Minister in the story. One in seven respondents believe that Bishop Zoltán Balog was instrumental in the decision being made.</p>
    <p> Novák Katalin elnök, aki K. Endre elnöki kegyelmet adott, csak négyből egynek van véleménye az elnök és a miniszterelnök szerepéről. Hétből egy válaszadó úgy véli, hogy Balog Zoltán püspök fontos szerepet játszott a meghozott döntésben.</p>
  </li>
  
  <li>
    <h2>Orbán met with Trump's former senior adviser in Washington D.C.</h2>
    <p><a href="https://telex.hu/english/2024/03/07/orban-met-with-trumps-former-senior-adviser-in-washington-d-c">https://telex.hu/english/2024/03/07/orban-met-with-trumps-former-senior-adviser-in-washington-d-c</a></p>
    <p>Prime Minister Viktor Orbán spoke on Thursday at a panel discussion about US-Hungarian relations in Washington, DC. The Heritage Foundation has been a major think tank of the American right for decades, and according to Kevin Roberts, its new goal is to institutionalise Trumpism. The audience included some of the most prominent American right-wing politicians, analysts and public figures.</p>
    <p>Orbán Viktor miniszterelnök csütörtökön a washingtoni (DC) amerikai-magyar kapcsolatokról tartott megbeszélésen beszélt. A Heritage Foundation évtizedek óta az amerikai jog egyik fő agytrösztje, és Kevin Roberts szerint új célja a Trumpizmus intézményesítése. A közönség a legkiemelkedőbb amerikai jobboldali politikusokat, elemzőket és közszereplőket is magában foglalta.</p>
  </li>
  
  <li>
    <h2>UNESCO warns Hungary against implementing original project planned for Lake Fertő</h2>
    <p><a href="https://telex.hu/english/2024/03/07/unesco-warns-hungary-against-implementing-the-original-project-planned-for-lake-ferto">https://telex.hu/english/2024/03/07/unesco-warns-hungary-against-implementing-the-original-project-planned-for-lake-ferto</a></p>
    <p>Lake Fertő-Neusiedlersee Cultural Landscape has been a World Heritage Site since 2001. UNESCO's team was joined by experts from ICOMOS, the international organisation for the conservation and protection of cultural heritage sites. Their findings show that the lake is already under pressure from development, tourism, hydrological changes as well as the impact of climate change.</p>
    <p> Lake Fertő-Neusiedlersee A kulturális táj 2001 óta a világörökség része. Az UNESCO csapatát az ICOMOS, a kulturális örökség megőrzéséért és védelméért felelős nemzetközi szervezet szakemberei is csatlakoztak. Az eredmények azt mutatják, hogy a tónak már a fejlődés, az idegenforgalom, a hidrológiai változások, valamint az éghajlatváltozás hatása is nyomás alatt áll.</p>
  </li>
  
</ul>
</body>
</html>


# A megoldás:

## Feltelepítjük a megfelelő modulokat.

In [67]:
!pip install newspaper3k
!pip install googletrans==4.0.0-rc1
!pip install transformers

## Kiszedjük az arhivum cikkeinek összes (nem teljes!) url-jét az első oldalról.

In [68]:
# Létrehozunk egy üres listát az url-ek tárolására
urls = []
start = 1
end = 10 # jelenleg 7539 a legnagyobb lehetséges érték

# Végigmegyünk a dátumokon egy napos lépésekkel, amíg el nem érjük a végdátumot
while start < end:
    # Formázzuk a dátumot úgy, hogy két számjegyű legyen a hónap és a nap
    date_str = f"{start}"
    # Összerakjuk az url-t a dátummal
    url = f"https://telex.hu/english?oldal=1={date_str}"
    # Hozzáadjuk az url-t a listához
    urls.append(url)
    # Növeljük a dátumot egy nappal
    start += 1

# Kiírjuk az url-ek listáját
print(urls)

['https://telex.hu/english?oldal=1=1', 'https://telex.hu/english?oldal=1=2', 'https://telex.hu/english?oldal=1=3', 'https://telex.hu/english?oldal=1=4', 'https://telex.hu/english?oldal=1=5', 'https://telex.hu/english?oldal=1=6', 'https://telex.hu/english?oldal=1=7', 'https://telex.hu/english?oldal=1=8', 'https://telex.hu/english?oldal=1=9']


## Beszerezzük a cikkek (teljes!!!) URL linkjét és címeit.

In [69]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

url = 'https://telex.hu/english?oldal=1'
news = []

try:
    response = requests.get(url)
    response.raise_for_status()
except requests.exceptions.RequestException as e:
    print(e)
    sys.exit(1)

soup = BeautifulSoup(response.content, 'html.parser')

articles = soup.find_all('a', class_="list__item__title", href=True)

df = pd.DataFrame(columns=['Title', 'Link', 'Text'])

for article in articles[:10]:  # Az első 10 cikk feldolgozása
    title = article.find(class_="hasHighlight").text.strip()
    link = f'https://telex.hu' + article["href"]
    news.append({"Title": title, "Link": link})

df = pd.DataFrame(news)
df.to_csv("telex_english_news.csv", index=False)


## Kiíratjuk az előzőeket ellenőrzésképp.

In [70]:
import pandas as pd

# Adatok betöltése DataFrame-be
df = pd.read_csv("telex_english_news.csv")

# DataFrame tartalmának kiírása
print(df)


                                               Title  \
0                           Arte: My body, my rules?   
1  Hungarian FM summons US ambassador over Biden’...   
2  Hungarian government only one to vote against ...   
3  Trump-admiring fake Twitter profile delivers p...   
4                           Orbán: Trump is a legend   
5            Set decorator Zsuzsa Mihalek wins Oscar   
6  He's the boss, he's a great leader – Trump hos...   
7  VSquare: Tusk and Fiala were shouting at Orbán...   
8  Majority holds Orbán responsible for clemency ...   
9  Orbán met with Trump's former senior adviser i...   

                                                Link  
0  https://telex.hu/english/2024/03/13/arte-abort...  
1  https://telex.hu/english/2024/03/12/hungarian-...  
2  https://telex.hu/english/2024/03/12/hungarian-...  
3  https://telex.hu/english/2024/03/12/trump-admi...  
4  https://telex.hu/english/2024/03/11/orban-trum...  
5  https://telex.hu/english/2024/03/11/set-decora... 

## Az előzőek ismeretében már le tudjuk tölteni a tartalmat:

In [71]:
import pandas as pd
from newspaper import Article

# Adatok betöltése DataFrame-be
df = pd.read_csv("telex_english_news.csv")

# Cikkek tartalmának letöltése
for index, row in df.iterrows():
    link = row['Link']
    article = Article(link)
    article.download()
    article.parse()
    text = article.text
    # A cikk tartalmának hozzáadása a DataFrame-hez
    df.at[index, 'Text'] = text

df.to_csv("telex_news_with_summary.csv", index=False)

# DataFrame tartalmának kiírása
print(df)
print(len(df))
#print(df.loc[9, 'Text'])

                                               Title  \
0                           Arte: My body, my rules?   
1  Hungarian FM summons US ambassador over Biden’...   
2  Hungarian government only one to vote against ...   
3  Trump-admiring fake Twitter profile delivers p...   
4                           Orbán: Trump is a legend   
5            Set decorator Zsuzsa Mihalek wins Oscar   
6  He's the boss, he's a great leader – Trump hos...   
7  VSquare: Tusk and Fiala were shouting at Orbán...   
8  Majority holds Orbán responsible for clemency ...   
9  Orbán met with Trump's former senior adviser i...   

                                                Link  \
0  https://telex.hu/english/2024/03/13/arte-abort...   
1  https://telex.hu/english/2024/03/12/hungarian-...   
2  https://telex.hu/english/2024/03/12/hungarian-...   
3  https://telex.hu/english/2024/03/12/trump-admi...   
4  https://telex.hu/english/2024/03/11/orban-trum...   
5  https://telex.hu/english/2024/03/11/set-deco

## A tartalmak között lehetnek hibás struktúrájú adatok is, mi most ezeket eldobjuk.

In [72]:
trashLine = 0

for index, row in df.iterrows():
  print(index,row['Text'][0:20]+"\n")
  if str(row['Text']) == "":
    trashLine = index
    df = df.drop(index=trashLine)

df.to_csv("telex_news_with_summary.csv", index=False)
df = pd.read_csv("telex_news_with_summary.csv")

print(df)
print(len(df))

0 France has become th

1 "Hungary doesn't hav

2 Member states voted 

3 The perfect case for

4 "He is indeed the se

5 

6 Donald Trump and a s

7 Prime Minister Vikto

8 A representative pol

9 Prime Minister Vikto

                                               Title  \
0                           Arte: My body, my rules?   
1  Hungarian FM summons US ambassador over Biden’...   
2  Hungarian government only one to vote against ...   
3  Trump-admiring fake Twitter profile delivers p...   
4                           Orbán: Trump is a legend   
5  He's the boss, he's a great leader – Trump hos...   
6  VSquare: Tusk and Fiala were shouting at Orbán...   
7  Majority holds Orbán responsible for clemency ...   
8  Orbán met with Trump's former senior adviser i...   

                                                Link  \
0  https://telex.hu/english/2024/03/13/arte-abort...   
1  https://telex.hu/english/2024/03/12/hungarian-...   
2  https://telex.hu/english/2024/03/12/hungarian-..

# Elkészítjük a modelt.

In [73]:
import pandas as pd
from newspaper import Article
from transformers import BartTokenizer, TFBartForConditionalGeneration
from transformers import pipeline
from googletrans import Translator

# Adatok betöltése DataFrame-be
df = pd.read_csv("telex_news_with_summary.csv")

# BART model betöltése
tokenizer = BartTokenizer.from_pretrained('facebook/bart-large-cnn')
model = TFBartForConditionalGeneration.from_pretrained('facebook/bart-large-cnn')

# Érzelmi analízis létrehozása
sentiment_classifier = pipeline("sentiment-analysis", model="cardiffnlp/twitter-roberta-base-sentiment-latest")

# Fordítás angolról magyarra
translator = Translator()

# Összefoglalás létrehozása
def generate_summary(text):
    if isinstance(text, str):
        text = [text]  # Ha csak egy stringet kapunk, alakítsuk át listává
    inputs = tokenizer(text, max_length=1024, return_tensors='tf', truncation=True, padding=True)
    summary_ids = model.generate(inputs['input_ids'], max_length=80, num_beams=4, early_stopping=True)
    summaries = [tokenizer.decode(summary_id, skip_special_tokens=True) for summary_id in summary_ids]
    return summaries[0] if len(summaries) == 1 else summaries  # Ha több összefoglalást generálunk, csak az elsőt adjuk vissza

# Érzelmi hangulat értékelése
def evaluate_sentiment(summary):
    result = sentiment_classifier(summary)
    return result[0]

# Fordítás angolról magyarra
def translate_to_hungarian(summary):
    translation = translator.translate(summary, src='en', dest='hu')
    return translation.text

# Összefoglalások, érzelmi hangulat és fordítás hozzáadása a DataFrame-hez
df['Summarized'] = df['Text'].apply(generate_summary)
df['Sentiment'] = df['Summarized'].apply(evaluate_sentiment)
df['Translated_Summary'] = df['Summarized'].apply(translate_to_hungarian)

# Emailhez előkészített DataFrame létrehozása HTML formátumban
email_df = df[['Title', 'Link', 'Summarized', 'Sentiment', 'Translated_Summary']]
email_html = email_df.to_html(index=False)

All PyTorch model weights were used when initializing TFBartForConditionalGeneration.

All the weights of TFBartForConditionalGeneration were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBartForConditionalGeneration for predictions without further training.
Some weights of the model checkpoint at cardiffnlp/twitter-roberta-base-sentiment-latest were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequence

## Az elkészült modell kiemetét megjelenítjük!
### Ez még csak egy nyers HTML táblázat, át kell alakítanunk!

In [75]:
print(email_html)

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Title</th>
      <th>Link</th>
      <th>Summarized</th>
      <th>Sentiment</th>
      <th>Translated_Summary</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Arte: My body, my rules?</td>
      <td>https://telex.hu/english/2024/03/13/arte-abortion-rights-france-slovakia</td>
      <td>France has become the first country in the world to enshrine the right to abortion in its constitution. Could other EU countries follow suit? Arte takes a closer look at the situation across Europe. This weekly News programme is produced by the Franco-German broadcaster ARTE and made available in 8 languages through a European media partnership.</td>
      <td>{'label': 'neutral', 'score': 0.7897151708602905}</td>
      <td>Franciaország lett az első ország a világon, amely az alkotmányában az abortuszhoz való jogot rögzítette.Más uniós országok követhetik -e a példát?Arte közelebbről megvizsgálja a helyze

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Title</th>
      <th>Link</th>
      <th>Summarized</th>
      <th>Sentiment</th>
      <th>Translated_Summary</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Arte: My body, my rules?</td>
      <td>https://telex.hu/english/2024/03/13/arte-abortion-rights-france-slovakia</td>
      <td>France has become the first country in the world to enshrine the right to abortion in its constitution. Could other EU countries follow suit? Arte takes a closer look at the situation across Europe. This weekly News programme is produced by the Franco-German broadcaster ARTE and made available in 8 languages through a European media partnership.</td>
      <td>{'label': 'neutral', 'score': 0.7897151708602905}</td>
      <td>Franciaország lett az első ország a világon, amely az alkotmányában az abortuszhoz való jogot rögzítette.Más uniós országok követhetik -e a példát?Arte közelebbről megvizsgálja a helyzetet Európában.Ezt a heti hírprogramot az Arte Franco-German Broadcaster készítette, és 8 nyelven elérhetővé teszi egy európai média partnerség útján.</td>
    </tr>
    <tr>
      <td>Hungarian FM summons US ambassador over Biden’s remarks about Orbán</td>
      <td>https://telex.hu/english/2024/03/12/hungarian-fm-summons-us-ambassador-over-bidens-remarks-about-orban</td>
      <td>US President Joe Biden criticised Hungarian Prime Minister Viktor Orbán. The Foreign Ministry summoned the US ambassador for an "urgent" meeting on Tuesday. The ambassador was received by Deputy Minister Levente Magyar, who asked him to produce the quote with the place and time where Orbán had said it.</td>
      <td>{'label': 'neutral', 'score': 0.6591066122055054}</td>
      <td>Joe Biden amerikai elnök kritizálta Viktor Orbán magyar miniszterelnököt.A külügyminisztérium kedden meghívta az Egyesült Államok nagykövetét egy "sürgős" ülésre.A nagykövet Levente Magyar miniszter -helyettes fogadta, aki arra kérte, hogy készítse el az árajánlatot azzal a helygel és idővel, ahol Orbán ezt mondta.</td>
    </tr>
    <tr>
      <td>Hungarian government only one to vote against EU law on transparency in political advertising</td>
      <td>https://telex.hu/english/2024/03/12/hungarian-government-only-one-to-vote-against-eu-law-on-transparency-in-political-advertising</td>
      <td>Member states voted on a regulation on transparency in political advertising in the EU Council on Monday. The Hungarian government's representative was the only one to vote against the regulation, while two delegates – from Austria and Estonia – abstained. A qualified majority (i.e. 15 out of 27 countries accounting for 65% of the EU's total population) was enough for the proposal to pass.</td>
      <td>{'label': 'neutral', 'score': 0.8929519653320312}</td>
      <td>A tagállamok hétfőn szavaztak az EU Tanácsának politikai reklámozásának átláthatóságáról szóló rendeletről.A magyar kormány képviselője volt az egyetlen, aki szavazott a rendelet ellen, míg két küldött - Ausztriából és Észtországból - tartózkodott.Az EU teljes népességének 65% -át teszi ki a minősített többség (azaz a 27 országból 15 -nél) elegendő volt a javaslat elfogadásához.</td>
    </tr>
    <tr>
      <td>Trump-admiring fake Twitter profile delivers perfect case for Sovereignty Protection Authority</td>
      <td>https://telex.hu/english/2024/03/12/trump-admiring-fake-twitter-profile-delivers-perfect-case-for-sovereignty-protection-authority</td>
      <td>The government-affiliated press in Hungary has been announcing in unison that there is now definite proof that George Soros funded the Hungarian opposition during the 2022 election campaign. The articles are referencing videos published on an X (formerly Twitter) account dedicated to praising Donald Trump under the name MagaBabe. The videos feature author Kati Marton, chair of A4D's New York advisory board</td>
      <td>{'label': 'neutral', 'score': 0.768927276134491}</td>
      <td>A magyarországi kormányhoz kapcsolódó sajtó egyhangúan bejelentette, hogy jelenleg határozott bizonyíték van arra, hogy George Soros a 2022-es választási kampány során finanszírozta a magyar ellenzéket.A cikkek egy X (korábban Twitter) beszámolóban közzétett videókra hivatkoznak, amelyek Donald Trump dicséretére szolgálnak Masabe néven.A videókban szereplő szerző, Kati Marton, az A4D New York -i tanácsadó testületének elnöke</td>
    </tr>
    <tr>
      <td>Orbán: Trump is a legend</td>
      <td>https://telex.hu/english/2024/03/11/orban-trump-is-a-legend</td>
      <td>Prime Minister Viktor Orbán gave an interview on public television on Sunday evening. He was talking about the inauguration of Hungary's new president, Tamás Sulyok. Orbán also spoke about his visit to Donald Trump's estate in Mar-a-Lago, Florida. He said that he and Trump are both pro-peace.</td>
      <td>{'label': 'neutral', 'score': 0.761019766330719}</td>
      <td>Orbán Viktor miniszterelnök vasárnap este interjút készített a nyilvános televízióban.Beszélt a magyar új elnök, Tamás Sulyok beiktatásáról.Orbán szintén beszélt a Donald Trump birtokában tett látogatásáról a floridai Mar-a-Lago-ban.Azt mondta, hogy ő és Trump egyaránt a Peace-t támogatják.</td>
    </tr>
    <tr>
      <td>He's the boss, he's a great leader – Trump hosts Orbán in friendly meeting at his Florida estate</td>
      <td>https://telex.hu/english/2024/03/09/hes-the-boss-hes-a-great-leader-trump-hosts-orban-in-friendly-meeting-at-his-florida-estate</td>
      <td>Donald Trump met with Hungarian Prime Minister Viktor Orbán at Mar-a-Lago. Orbán said he and Donald Trump were "peacemakers" CNN sources described the meeting as a friendly chat without an agenda. This was their third meeting, with their first meeting taking place on 13 May 2019 at the White House.</td>
      <td>{'label': 'neutral', 'score': 0.8897312879562378}</td>
      <td>Donald Trump találkozott Viktor Orbán magyar miniszterelnökkel a Mar-a-Lago-ban.Orbán azt mondta, hogy ő és Donald Trump "Peacemaker" CNN forrásai, a találkozót barátságos csevegésként írták le napirend nélkül.Ez volt a harmadik ülésük, amikor az első találkozójuk 2019. május 13 -án zajlott a Fehér Házban.</td>
    </tr>
    <tr>
      <td>VSquare: Tusk and Fiala were shouting at Orbán while Fico remained silent</td>
      <td>https://telex.hu/english/2024/03/08/vsquare-tusk-and-fiala-were-shouting-at-orban-while-fico-remained-silent</td>
      <td>Prime Minister Viktor Orbán of Hungary will host the V4 summit in Prague on 27 February 2024. The summit will be attended by the leaders of the Czech Republic, Poland, Slovakia and Poland. Orbán will be accompanied by the Prime Minister of Poland, Donald Tusk and Robert Fico.</td>
      <td>{'label': 'neutral', 'score': 0.8464767932891846}</td>
      <td>A magyarországi Viktor Orbán miniszterelnök 2024. február 27 -én fogadja a V4 csúcstalálkozót.Orbánot Lengyelország miniszterelnöke, Donald Tusk és Robert Fico kíséri.</td>
    </tr>
    <tr>
      <td>Majority holds Orbán responsible for clemency decision, survey shows</td>
      <td>https://telex.hu/english/2024/03/08/majority-holds-orban-responsible-for-clemency-decision-survey-shows</td>
      <td>Only one in four people are of the opinion that President Katalin Novák, who granted a presidential pardon to Endre K., is responsible. Opposition and governing party voters have very different views on the role of the President and the Prime Minister in the story, the survey reveals.</td>
      <td>{'label': 'neutral', 'score': 0.7048771977424622}</td>
      <td>A felelős Katalin Novák elnök, aki elnöki bocsánatot kapott Katalin elnöknek, csak egy negyedik embernek felel meg.Az ellenzék és az irányító párt szavazók nagyon eltérő véleményt képviselnek az elnök és a miniszterelnök szerepéről a történetben - mutatja ki a felmérés.</td>
    </tr>
    <tr>
      <td>Orbán met with Trump's former senior adviser in Washington D.C.</td>
      <td>https://telex.hu/english/2024/03/07/orban-met-with-trumps-former-senior-adviser-in-washington-d-c</td>
      <td>Prime Minister Viktor Orbán spoke on Thursday at a panel discussion about US-Hungarian relations hosted by the conservative-Trumpist Heritage Foundation in Washington, DC. The Heritage Foundation has been a major think tank of the American right for decades, and according to Kevin Roberts, its new goal is to institutionalise Trumpism. As the article of 444 recalls, the foundation is campaigning against helping</td>
      <td>{'label': 'neutral', 'score': 0.8596919775009155}</td>
      <td>Orbán Viktor miniszterelnök csütörtökön beszélt az amerikai magyar kapcsolatokról szóló panelbeszélgetésen, amelyet a Washington DC-ben a Konzervatív-Trumpista Örökség Alapítvány szervezett.A Heritage Alapítvány évtizedek óta az amerikai jog egyik fő gondolkodóközpontja, és Kevin Roberts szerint új célja a Trumpizmus intézményesítése.Ahogy a 444 -es cikk emlékeztet, az alapítvány a segítségnyújtás ellen kampányozik</td>
    </tr>
  </tbody>
</table>



## Az átalakítást megtéve:

In [106]:
import pandas as pd

# DataFrame létrehozása
df = pd.DataFrame(data)

emailList = []
# E-mail szövegek létrehozása HTML formátumban
for i, row in df.iterrows():
    email_html = f"""
    <b>Cím:</b> {row['Title']}<br>
    <b>Elérhetőség:</b> {row['Link']}<br>
    <b>Angol nyelvű összefoglaló:</b> {row['Summarized']}<br>
    <b>Hangulat:</b> {row['Sentiment']}<br>
    <b>Magyar nyelvű összefoglaló:</b> {row['Translated_Summary']}<br><br>"""
    emailList.append(email_html.strip())

## Az elkészült tartalom a következő:

In [107]:
for email in emailList:
  print(email)

<b>Cím:</b> Arte: My body, my rules?<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/13/arte-abortion-rights-france-slovakia<br>
    <b>Angol nyelvű összefoglaló:</b> France has become the first country in the world to enshrine the right to abortion in its constitution. Could other EU countries follow suit? Arte takes a closer look at the situation across Europe. This weekly News programme is produced by the Franco-German broadcaster ARTE and made available in 8 languages through a European media partnership.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.7897151708602905}<br>
    <b>Magyar nyelvű összefoglaló:</b> Franciaország lett az első ország a világon, amely az alkotmányában az abortuszhoz való jogot rögzítette.Más uniós országok követhetik -e a példát?Arte közelebbről megvizsgálja a helyzetet Európában.Ezt a heti hírprogramot az Arte Franco-German Broadcaster készítette, és 8 nyelven elérhetővé teszi egy európai média partnerség útján.<br><br>
<b>Cím:</b>

# Legvégül az előző kódot kimásoltam egy szövegblokkba, hogy lássuk a formázást is:



<b>Cím:</b> Arte: My body, my rules?<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/13/arte-abortion-rights-france-slovakia<br>
    <b>Angol nyelvű összefoglaló:</b> France has become the first country in the world to enshrine the right to abortion in its constitution. Could other EU countries follow suit? Arte takes a closer look at the situation across Europe. This weekly News programme is produced by the Franco-German broadcaster ARTE and made available in 8 languages through a European media partnership.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.7897151708602905}<br>
    <b>Magyar nyelvű összefoglaló:</b> Franciaország lett az első ország a világon, amely az alkotmányában az abortuszhoz való jogot rögzítette.Más uniós országok követhetik -e a példát?Arte közelebbről megvizsgálja a helyzetet Európában.Ezt a heti hírprogramot az Arte Franco-German Broadcaster készítette, és 8 nyelven elérhetővé teszi egy európai média partnerség útján.<br><br>
<b>Cím:</b> Hungarian FM summons US ambassador over Biden’s remarks about Orbán<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/12/hungarian-fm-summons-us-ambassador-over-bidens-remarks-about-orban<br>
    <b>Angol nyelvű összefoglaló:</b> US President Joe Biden criticised Hungarian Prime Minister Viktor Orbán. The Foreign Ministry summoned the US ambassador for an 'urgent' meeting on Tuesday. The ambassador was received by Deputy Minister Levente Magyar, who asked him to produce the quote with the place and time where Orbán had said it.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.6591066122055054}<br>
    <b>Magyar nyelvű összefoglaló:</b> Joe Biden amerikai elnök kritizálta Viktor Orbán magyar miniszterelnököt.A külügyminisztérium kedden meghívta az Egyesült Államok nagykövetét egy 'sürgős' ülésre.A nagykövet Levente Magyar miniszter -helyettes fogadta, aki arra kérte, hogy készítse el az árajánlatot azzal a helygel és idővel, ahol Orbán ezt mondta.<br><br>
<b>Cím:</b> Hungarian government only one to vote against EU law on transparency in political advertising<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/12/hungarian-government-only-one-to-vote-against-eu-law-on-transparency-in-political-advertising<br>
    <b>Angol nyelvű összefoglaló:</b> Member states voted on a regulation on transparency in political advertising in the EU Council on Monday. The Hungarian government's representative was the only one to vote against the regulation, while two delegates – from Austria and Estonia – abstained. A qualified majority (i.e. 15 out of 27 countries accounting for 65% of the EU's total population) was enough for the proposal to pass.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.8929519653320312}<br>
    <b>Magyar nyelvű összefoglaló:</b> A tagállamok hétfőn szavaztak az EU Tanácsának politikai reklámozásának átláthatóságáról szóló rendeletről.A magyar kormány képviselője volt az egyetlen, aki szavazott a rendelet ellen, míg két küldött - Ausztriából és Észtországból - tartózkodott.Az EU teljes népességének 65% -át teszi ki a minősített többség (azaz a 27 országból 15 -nél) elegendő volt a javaslat elfogadásához.<br><br>
<b>Cím:</b> Trump-admiring fake Twitter profile delivers perfect case for Sovereignty Protection Authority<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/12/trump-admiring-fake-twitter-profile-delivers-perfect-case-for-sovereignty-protection-authority<br>
    <b>Angol nyelvű összefoglaló:</b> The government-affiliated press in Hungary has been announcing in unison that there is now definite proof that George Soros funded the Hungarian opposition during the 2022 election campaign. The articles are referencing videos published on an X (formerly Twitter) account dedicated to praising Donald Trump under the name MagaBabe. The videos feature author Kati Marton, chair of A4D's New York advisory board.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.768927276134491}<br>
    <b>Magyar nyelvű összefoglaló:</b> A magyarországi kormányhoz kapcsolódó sajtó egyhangúan bejelentette, hogy jelenleg határozott bizonyíték van arra, hogy George Soros a 2022-es választási kampány során finanszírozta a magyar ellenzéket.A cikkek egy X (korábban Twitter) beszámolóban közzétett videókra hivatkoznak, amelyek Donald Trump dicséretére szolgálnak Masabe néven.A videókban szereplő szerző, Kati Marton, az A4D New York -i tanácsadó testületének elnöke<br><br>
<b>Cím:</b> Orbán: Trump is a legend<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/11/orban-trump-is-a-legend<br>
    <b>Angol nyelvű összefoglaló:</b> Prime Minister Viktor Orbán gave an interview on public television on Sunday evening. He was talking about the inauguration of Hungary's new president, Tamás Sulyok. Orbán also spoke about his visit to Donald Trump's estate in Mar-a-Lago, Florida. He said that he and Trump are both pro-peace.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.761019766330719}<br>
    <b>Magyar nyelvű összefoglaló:</b> Orbán Viktor miniszterelnök vasárnap este interjút készített a nyilvános televízióban.Beszélt a magyar új elnök, Tamás Sulyok beiktatásáról.Orbán szintén beszélt a Donald Trump birtokában tett látogatásáról a floridai Mar-a-Lago-ban.Azt mondta, hogy ő és Trump egyaránt a Peace-t támogatják.<br><br>
<b>Cím:</b> He's the boss, he's a great leader – Trump hosts Orbán in friendly meeting at his Florida estate<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/09/hes-the-boss-hes-a-great-leader-trump-hosts-orban-in-friendly-meeting-at-his-florida-estate<br>
    <b>Angol nyelvű összefoglaló:</b> Donald Trump met with Hungarian Prime Minister Viktor Orbán at Mar-a-Lago. Orbán said he and Donald Trump were 'peacemakers' CNN sources described the meeting as a friendly chat without an agenda. This was their third meeting, with their first meeting taking place on 13 May 2019 at the White House.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.8897312879562378}<br>
    <b>Magyar nyelvű összefoglaló:</b> Donald Trump találkozott Viktor Orbán magyar miniszterelnökkel a Mar-a-Lago-ban.Orbán azt mondta, hogy ő és Donald Trump 'Peacemaker' CNN forrásai, a találkozót barátságos csevegésként írták le napirend nélkül.Ez volt a harmadik ülésük, amikor az első találkozójuk 2019. május 13 -án zajlott a Fehér Házban.<br><br>
<b>Cím:</b> VSquare: Tusk and Fiala were shouting at Orbán while Fico remained silent<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/08/vsquare-tusk-and-fiala-were-shouting-at-orban-while-fico-remained-silent<br>
    <b>Angol nyelvű összefoglaló:</b> Prime Minister Viktor Orbán of Hungary will host the V4 summit in Prague on 27 February 2024. The summit will be attended by the leaders of the Czech Republic, Poland, Slovakia and Poland. Orbán will be accompanied by the Prime Minister of Poland, Donald Tusk and Robert Fico.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.8464767932891846}<br>
    <b>Magyar nyelvű összefoglaló:</b> A magyarországi Viktor Orbán miniszterelnök 2024. február 27 -én fogadja a V4 csúcstalálkozót.Orbánot Lengyelország miniszterelnöke, Donald Tusk és Robert Fico kíséri.<br><br>
<b>Cím:</b> Majority holds Orbán responsible for clemency decision, survey shows<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/08/majority-holds-orban-responsible-for-clemency-decision-survey-shows<br>
    <b>Angol nyelvű összefoglaló:</b> Only one in four people are of the opinion that President Katalin Novák, who granted a presidential pardon to Endre K., is responsible. Opposition and governing party voters have very different views on the role of the President and the Prime Minister in the story, the survey reveals.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.7048771977424622}<br>
    <b>Magyar nyelvű összefoglaló:</b> A felelős Katalin Novák elnök, aki elnöki bocsánatot kapott Katalin elnöknek, csak egy negyedik embernek felel meg.Az ellenzék és az irányító párt szavazók nagyon eltérő véleményt képviselnek az elnök és a miniszterelnök szerepéről a történetben - mutatja ki a felmérés.<br><br>
<b>Cím:</b> Orbán met with Trump's former senior adviser in Washington D.C.<br>
    <b>Elérhetőség:</b> https://telex.hu/english/2024/03/07/orban-met-with-trumps-former-senior-adviser-in-washington-d-c<br>
    <b>Angol nyelvű összefoglaló:</b> Prime Minister Viktor Orbán spoke on Thursday at a panel discussion about US-Hungarian relations hosted by the conservative-Trumpist Heritage Foundation in Washington, DC. The Heritage Foundation has been a major think tank of the American right for decades, and according to Kevin Roberts, its new goal is to institutionalise Trumpism. As the article of 444 recalls, the foundation is campaigning against helping.<br>
    <b>Hangulat:</b> {'label': 'neutral', 'score': 0.8596919775009155}<br>
    <b>Magyar nyelvű összefoglaló:</b> Orbán Viktor miniszterelnök csütörtökön beszélt az amerikai magyar kapcsolatokról szóló panelbeszélgetésen, amelyet a Washington DC-ben a Konzervatív-Trumpista Örökség Alapítvány szervezett.A Heritage Alapítvány évtizedek óta az amerikai jog egyik fő gondolkodóközpontja, és Kevin Roberts szerint új célja a Trumpizmus intézményesítése.Ahogy a 444 -es cikk emlékeztet, az alapítvány a segítségnyújtás ellen kampányozik.<br><br>

